README.md
# bottom_indicator_bar_svg

A flutter bottom tab with indicator, similar to the bottom tab of facebook app

## Forked Mods

- Allow svg icon assets for the icons

- Add labels to nav items

- Full credit to original repo by [Juan Jose Carracedo](https://github.com/DonsWayo)

![ezgif-2-f7cff4c02f](https://user-images.githubusercontent.com/4716711/150628594-1f1e96db-4bec-4809-977f-245a6fc546fc.gif "bottom_indicator_bar_svg")

## Getting Started

Add the dependency in `pubspec.yaml`:

```yaml
dependencies:
  ...
  bottom_indicator_bar_svg: lastest_version
```

## Basic Usage


```dart
class HomePage extends StatefulWidget {
  @override
  State createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  final List<BottomIndicatorNavigationBarItem> items = [
    BottomIndicatorNavigationBarItem(icon: Icons.home, label: Text('Home')),
    BottomIndicatorNavigationBarItem(icon: Icons.search, label: 'Search'),
    BottomIndicatorNavigationBarItem(icon: 'assets/svgIcon.svg', label: 'Svg', iconSize: 28),
  ];


  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Indicator Bottom Bar"),
        backgroundColor: Colors.black87,
      ),
      body: Center(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
          ],
        ),
      ),
      bottomNavigationBar: BottomIndicatorBar(
        onTap: (index) => {},
        items: items,
        iconSize: 30.0,
        indicatorHeight: 5, // Set to 0 to hide the indicator bar
        activeColor: Colors.blue,
        inactiveColor: Colors.grey,
        indicatorColor: Colors.blue,
        backgroundColor: Colors.black87,
      ),
    );
  }
}
```


CHANGELOG.md
## [1.0.0] - First Release.

* First release

## [2.0.0] - 2021/06/11.

* Upgrade null safety and update example project 

## [2.1.0] - 2022/01/21.

* Add ability to use svg images as icons 

## [2.1.2] - 2022/01/21.

* Add ability to add labels 

## [2.1.3] - 2022/01/22.

* Add ability to customize icon size and fix label selection color 

## [2.1.5] - 2022/01/22.

* Add ability to set active svg icon and ability to customize bar height

## [3.0.0] - 2022/10/28.

* Add ability to set the indicator bar height. Set to 0 to hide.
* Add icon size override onto BottomIndicatorNavigationBarItem

lib/src/navigation_bar.dart
import 'dart:ui' show lerpDouble;

import 'package:flutter/material.dart';
import 'package:flutter_svg/svg.dart';

import 'navigation_bar_item.dart';

// ignore: must_be_immutable
class BottomIndicatorBar extends StatefulWidget {
  final Color indicatorColor;
  final Color activeColor;
  final Color inactiveColor;

  /// Specify the indicator bar height. (Set to 0 to hide the indicator bar)
  final double indicatorHeight;
  final Color backgroundColor;
  final bool shadow;
  final double iconSize;
  final double barHeight;
  int currentIndex;
  final ValueChanged<int> onTap;
  final List<BottomIndicatorNavigationBarItem> items;

  BottomIndicatorBar({
    Key? key,
    required this.onTap,
    required this.items,
    this.activeColor = Colors.teal,
    this.inactiveColor = Colors.grey,
    this.indicatorColor = Colors.grey,
    this.backgroundColor = Colors.white,
    this.indicatorHeight = 2,
    this.iconSize = 35.0,
    this.barHeight = 60,
    this.shadow = true,
    this.currentIndex = 0,
  }) : super(key: key);

  @override
  State createState() => _BottomIndicatorBarState();
}

class _BottomIndicatorBarState extends State<BottomIndicatorBar> {
  //static const double INDICATOR_HEIGHT = 2;

  List<BottomIndicatorNavigationBarItem> get items => widget.items;

  double width = 0;
  late Color activeColor;
  Duration duration = Duration(milliseconds: 170);

  double? _getIndicatorPosition(int index) {
    var isLtr = Directionality.of(context) == TextDirection.ltr;
    if (isLtr)
      return lerpDouble(-1.0, 1.0, index / (items.length - 1));
    else
      return lerpDouble(1.0, -1.0, index / (items.length - 1));
  }

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    width = MediaQuery.of(context).size.width;
    activeColor = widget.activeColor;

    return Container(
      height: widget.barHeight + MediaQuery.of(context).viewPadding.bottom,
      width: width,
      decoration: BoxDecoration(
        color: widget.backgroundColor,
        boxShadow: widget.shadow
            ? [
                BoxShadow(color: Colors.black12, blurRadius: 10),
              ]
            : null,
      ),
      child: Stack(
        clipBehavior: Clip.hardEdge,
        children: <Widget>[
          Positioned(
            top: widget.indicatorHeight,
            child: Row(
              crossAxisAlignment: CrossAxisAlignment.end,
              children: items.map((item) {
                var index = items.indexOf(item);
                return GestureDetector(
                  onTap: () => _select(index, item),
                  child: _buildItemWidget(item, index == widget.currentIndex),
                );
              }).toList(),
            ),
          ),
          Positioned(
            top: 0,
            width: width,
            child: AnimatedAlign(
              alignment:
                  Alignment(_getIndicatorPosition(widget.currentIndex)!, 0),
              curve: Curves.linear,
              duration: duration,
              child: Container(
                color: widget.indicatorColor,
                width: width / items.length,
                height: widget.indicatorHeight,
              ),
            ),
          ),
        ],
      ),
    );
  }

  _select(int index, BottomIndicatorNavigationBarItem item) {
    widget.currentIndex = index;
    widget.onTap(widget.currentIndex);

    setState(() {});
  }

  Widget _setIcon(BottomIndicatorNavigationBarItem item, bool isSelected) {
    if (item.icon is IconData) {
      return Icon(
        item.icon,
        color: isSelected ? activeColor : widget.inactiveColor,
        //size: widget.iconSize,
        size: item.iconSize != null ? item.iconSize : widget.iconSize,
      );
    } else if (item.icon is String && item.icon.contains('.svg')) {
      if (item.activeIcon != null && item.activeIcon is String) {
        var asset = item.icon;
        if (isSelected) {
          asset = item.activeIcon;
        }
        return SvgPicture.asset(
          asset,
          //height: widget.iconSize,
          height: item.iconSize != null ? item.iconSize : widget.iconSize,
        );
      }
      return SvgPicture.asset(
        item.icon,
        color: isSelected ? activeColor : widget.inactiveColor,
        height: item.iconSize != null ? item.iconSize : widget.iconSize,
      );
    } else {
      return item.icon;
    }
  }

  Widget _setLabel(BottomIndicatorNavigationBarItem item, bool isSelected) {
    if (item.label == null) {
      return SizedBox.shrink();
    }
    if (item.label is String) {
      return Padding(
        padding: const EdgeInsets.all(4.0),
        child: Text(
          item.label,
          style:
              TextStyle(color: isSelected ? activeColor : widget.inactiveColor),
        ),
      );
    }
    if (item.label is Text) {
      var textWidget = (item.label as Text);
      if (textWidget.style != null) {
        textWidget.style!.copyWith(color: Colors.white);
      } else {
        return Text(
          textWidget.data!,
          style:
              TextStyle(color: isSelected ? activeColor : widget.inactiveColor),
        );
      }
    }
    return item.label;
  }

  Widget _buildItemWidget(
      BottomIndicatorNavigationBarItem item, bool isSelected) {
    return Container(
      color: item.backgroundColor,
      height: widget.barHeight,
      width: width / items.length,
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          Column(
            children: [_setIcon(item, isSelected), _setLabel(item, isSelected)],
          ),
        ],
      ),
    );
  }
}


example/lib/main.dart
import 'package:bottom_indicator_bar_svg/bottom_indicator_bar_svg.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Titled Bar',
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  State createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  final List<BottomIndicatorNavigationBarItem> items = [
    BottomIndicatorNavigationBarItem(icon: Icons.home, label: Text('Home')),
    BottomIndicatorNavigationBarItem(icon: Icons.search, label: 'Search'),
    BottomIndicatorNavigationBarItem(icon: 'assets/svgIcon.svg', label: 'Svg'),
    BottomIndicatorNavigationBarItem(
        icon: 'assets/inactiveIcon.svg',
        activeIcon: 'assets/activeIcon.svg',
        label: 'Account',
        iconSize: 28),
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Indicator Bottom Bar"),
        backgroundColor: Colors.black87,
      ),
      body: Center(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[],
        ),
      ),
      bottomNavigationBar: BottomIndicatorBar(
        onTap: (index) => print("Selected Index: $index"),
        items: items,
        iconSize: 30.0,
        barHeight: 70.0,
        activeColor: Colors.blue,
        inactiveColor: Colors.grey,
        indicatorColor: Colors.blue,
        backgroundColor: Colors.black87,
        indicatorHeight: 5,
      ),
    );
  }
}
